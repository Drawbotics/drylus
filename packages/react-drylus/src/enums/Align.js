import Enum from '@drawbotics/enums';


const Align = new Enum(
  'LEFT',
  'CENTER',
  'RIGHT',
);


export default Align;
