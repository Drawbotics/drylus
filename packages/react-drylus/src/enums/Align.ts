export enum Align {
  LEFT = 'LEFT',
  CENTER = 'CENTER',
  RIGHT = 'RIGHT',
  JUSTIFY = 'JUSTIFY',
}
