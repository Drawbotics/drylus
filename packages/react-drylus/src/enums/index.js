export Category from './Category';
export Tier from './Tier';
export Size from './Size';
export Position from './Position';
export Align from './Align';
export Color from './Color';
export Shade from './Shade';
