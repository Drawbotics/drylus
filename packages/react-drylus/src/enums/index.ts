export * from './Category';
export * from './Tier';
export * from './Size';
export * from './Position';
export * from './Align';
export * from './Color';
export * from './Shade';
