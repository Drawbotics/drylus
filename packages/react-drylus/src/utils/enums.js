import camelCase from 'lodash/camelCase';

import { Category, Color, Shade, Tier } from '../enums';

export function getEnumAsClass(enumVal) {
  return camelCase(enumVal?.description?.toLowerCase());
}

export function colorEnumToCategory(enumVal) {
  switch (enumVal) {
    case Color.RED:
      return Category.DANGER;
    case Color.GREEN:
      return Category.SUCCESS;
    case Color.ORANGE:
      return Category.WARNING;
    case Color.BLUE:
      return Category.INFO;
  }
  return enumVal;
}

export function categoryEnumToColor(enumVal) {
  switch (enumVal) {
    case Category.DANGER:
      return Color.RED;
    case Category.SUCCESS:
      return Color.GREEN;
    case Category.WARNING:
      return Color.ORANGE;
    case Category.INFO:
      return Color.BLUE;
  }
  return enumVal;
}

export function shadeEnumToTier(enumVal) {
  switch (enumVal) {
    case Shade.DARK:
      return Tier.PRIMARY;
    case Shade.MEDIUM:
      return Tier.SECONDARY;
    case Shade.LIGHT:
      return Tier.TERTIARY;
  }
  return enumVal;
}

export function getIconForCategory(category) {
  switch (category) {
    case Category.DANGER:
      return 'alert-circle';
    case Category.SUCCESS:
      return 'check-circle';
    case Category.WARNING:
      return 'alert-triangle';
    default:
      return 'info';
  }
}
