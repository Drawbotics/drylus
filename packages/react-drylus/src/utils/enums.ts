import camelCase from 'lodash/camelCase';

import { Category, Color, Position, Shade, Size, Tier } from '../enums';

type Enum = Category | Color | Shade | Tier | Size | Position;

export function getEnumAsClass<T extends Record<string, string>>(enumVal?: Enum): keyof T {
  return camelCase(enumVal?.toLowerCase());
}

export function colorEnumToCategory(enumVal: Color): Category | undefined {
  switch (enumVal) {
    case Color.RED:
      return Category.DANGER;
    case Color.GREEN:
      return Category.SUCCESS;
    case Color.ORANGE:
      return Category.WARNING;
    case Color.BLUE:
      return Category.INFO;
  }
}

export function categoryEnumToColor(enumVal: Category): Color | undefined {
  switch (enumVal) {
    case Category.DANGER:
      return Color.RED;
    case Category.SUCCESS:
      return Color.GREEN;
    case Category.WARNING:
      return Color.ORANGE;
    case Category.INFO:
      return Color.BLUE;
  }
}

export function shadeEnumToTier(enumVal: Shade): Tier | undefined {
  switch (enumVal) {
    case Shade.DARK:
      return Tier.PRIMARY;
    case Shade.MEDIUM:
      return Tier.SECONDARY;
    case Shade.LIGHT:
      return Tier.TERTIARY;
  }
}

export function getIconForCategory(category: Category): string {
  switch (category) {
    case Category.DANGER:
      return 'alert-circle';
    case Category.SUCCESS:
      return 'check-circle';
    case Category.WARNING:
      return 'alert-triangle';
    default:
      return 'info';
  }
}
