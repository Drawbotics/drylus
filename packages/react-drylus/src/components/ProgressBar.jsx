import React from 'react';
import { css, cx, keyframes } from 'emotion';
import PropTypes from 'prop-types';
import sv from '@drawbotics/drylus-style-vars';

import { Categories, Sizes } from '../base';
import { getEnumAsClass } from '../utils';


const translateX = keyframes`
  0% {
    transform: translateX(-40%);
  }
  100% {
    transform: translateX(140%);
  }
`;


const styles = {
  root: css`
    height: ${sv.marginExtraSmall};
    width: 100%;
    position: relative;
    border-radius: ${sv.marginExtraSmall};
    overflow: hidden;
    background: ${sv.neutralLight};

    [data-element="bar"] {
      border-radius: ${sv.marginExtraSmall};
    }
  `,
  bar: css`
    position: absolute;
    z-index: 2;
    top: 0;
    left: 0;
    height: 100%;
    width: 0;
    background: ${sv.neutralDark};
    transition: ${sv.transitionShort};
    overflow: hidden;
  `,
  indeterminate: css`
    width: 50% !important;
    animation: ${translateX}
      calc(${sv.defaultTransitionTime} * 2)
      ease-in-out
      alternate
      infinite;
  `,
  large: css`
    height: ${sv.marginSmall};
    border-radius: ${sv.marginSmall};

    [data-element="bar"] {
      border-radius: ${sv.marginSmall};
    }
  `,
  extraLarge: css`
    height: 90px;
  `,
  small: css`
    height: calc(${sv.marginExtraSmall} / 2);
    border-radius: calc(${sv.marginExtraSmall} / 2);

    [data-element="bar"] {
      border-radius: calc(${sv.marginExtraSmall} / 2);
    }
  `,
  brand: css`
    background: ${sv.brand};
  `,
  danger: css`
    background: ${sv.red};
  `,
  info: css`
    background: ${sv.blue};
  `,
  warning: css`
    background: ${sv.orange};
  `,
  success: css`
    background: ${sv.green};
  `,
};


const ProgressBar = ({
  percentage,
  category,
  size,
  style,
}) => {
  const indeterminate = percentage == null;

  return (
    <div
      style={style}
      className={cx(styles.root, {
        [styles[getEnumAsClass(size)]]: size,
      })}>
      <div
        data-element="bar"
        className={cx(styles.bar, {
          [styles[getEnumAsClass(category)]]: category,
          [styles.indeterminate]: indeterminate,
        })}
        style={{ width: `${percentage * 100}%` }} />
    </div>
  );
};


ProgressBar.propTypes = {
  /** Determines the amount of the bar which is completed, between 0 and 1. If not given the bar is indeterminate */
  percentage: PropTypes.number,

  category: PropTypes.oneOf([
    Categories.BRAND,
    Categories.DANGER,
    Categories.SUCCESS,
    Categories.INFO,
    Categories.WARNING,
  ]),

  size: PropTypes.oneOf([
    Sizes.SMALL,
    Sizes.DEFAULT,
    Sizes.LARGE,
  ]),

  /** Used for style overrides */
  style: PropTypes.object,
};


ProgressBar.defaultProps = {
  size: Sizes.DEFAULT,
};


export default ProgressBar;
