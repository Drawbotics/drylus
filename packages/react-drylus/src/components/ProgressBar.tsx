import sv from '@drawbotics/drylus-style-vars';
import { css, cx } from 'emotion';
import React from 'react';

import { Color, Size } from '../enums';
import { Responsive, Style } from '../types';
import { getEnumAsClass, useResponsiveProps } from '../utils';
import { placeholderStyles as shimmerStyles } from './LoadingPlaceholder';

const styles = {
  root: css`
    height: ${sv.marginExtraSmall};
    width: 100%;
    position: relative;
    border-radius: ${sv.marginExtraSmall};
    overflow: hidden;
    background: ${sv.neutralLight};

    [data-element='bar'] {
      border-radius: ${sv.marginExtraSmall};
    }
  `,
  bar: css`
    position: absolute;
    z-index: 2;
    top: 0;
    left: 0;
    height: 100%;
    width: 0;
    background: ${sv.neutralDark};
    transition: ${sv.transitionShort};
    overflow: hidden;

    &::after {
      background: ${sv.neutralDark};
      border-radius: 0 !important;
      background-image: linear-gradient(
        to right,
        ${sv.neutralDark} 8%,
        ${sv.neutral} 18%,
        ${sv.neutralDark} 33%
      ) !important;
    }
  `,
  indeterminate: css`
    width: 100% !important;
  `,
  large: css`
    height: ${sv.marginSmall};
    border-radius: ${sv.marginSmall};

    [data-element='bar'] {
      border-radius: ${sv.marginSmall};
    }
  `,
  extraLarge: css`
    height: 90px;
  `,
  small: css`
    height: calc(${sv.marginExtraSmall} / 2);
    border-radius: calc(${sv.marginExtraSmall} / 2);

    [data-element='bar'] {
      border-radius: calc(${sv.marginExtraSmall} / 2);
    }
  `,
  brand: css`
    background: ${sv.brand};

    &::after {
      background: ${sv.brand};
      background-image: linear-gradient(
        to right,
        ${sv.brand} 8%,
        ${sv.brandLight} 18%,
        ${sv.brand} 33%
      ) !important;
    }
  `,
  red: css`
    background: ${sv.red};

    &::after {
      background: ${sv.red};
      background-image: linear-gradient(
        to right,
        ${sv.red} 8%,
        ${sv.redLight} 18%,
        ${sv.red} 33%
      ) !important;
    }
  `,
  blue: css`
    background: ${sv.blue};

    &::after {
      background: ${sv.blue};
      background-image: linear-gradient(
        to right,
        ${sv.blue} 8%,
        ${sv.blueLight} 18%,
        ${sv.blue} 33%
      ) !important;
    }
  `,
  orange: css`
    background: ${sv.orange};

    &::after {
      background: ${sv.orange};
      background-image: linear-gradient(
        to right,
        ${sv.orange} 8%,
        ${sv.orangeLight} 18%,
        ${sv.orange} 33%
      ) !important;
    }
  `,
  green: css`
    background: ${sv.green};

    &::after {
      background: ${sv.green};
      background-image: linear-gradient(
        to right,
        ${sv.green} 8%,
        ${sv.greenLight} 18%,
        ${sv.green} 33%
      ) !important;
    }
  `,
};

export interface ProgressBarProps {
  /** Determines the amount of the bar which is completed, between 0 and 1. If not given the bar is indeterminate */
  percentage?: number;

  /** @kind Color */
  color?: Color.BRAND | Color.RED | Color.BLUE | Color.GREEN | Color.ORANGE;

  /**
   * @default Size.DEFAULT
   * @kind Size
   */
  size?: Size.SMALL | Size.DEFAULT | Size.LARGE;

  /** Used for style overrides */
  style?: Style;

  /** Used for style overrides */
  className?: string;

  /** Reponsive prop overrides */
  responsive?: Responsive<this>;
}

export const ProgressBar = ({ responsive, ...rest }: ProgressBarProps) => {
  const { percentage, size = Size.DEFAULT, style, color, className } = useResponsiveProps<
    ProgressBarProps
  >(rest, responsive);

  const indeterminate = percentage == null;

  return (
    <div
      style={style}
      className={cx(
        styles.root,
        {
          [styles[getEnumAsClass<typeof styles>(size)]]: size != null,
        },
        className,
      )}>
      <div
        data-element="bar"
        className={cx(styles.bar, {
          [styles[getEnumAsClass<typeof styles>(color)]]: color != null,
          [styles.indeterminate]: indeterminate,
          [shimmerStyles.shimmer]: indeterminate,
        })}
        style={{ width: `${(percentage ?? 0) * 100}%` }}
      />
    </div>
  );
};
