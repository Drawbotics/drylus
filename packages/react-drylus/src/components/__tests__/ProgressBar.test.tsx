import React from 'react';
import { create } from 'react-test-renderer';

import { Color, Size } from '../../enums';
import { ProgressBar } from '../ProgressBar';

describe('ProgressBar', () => {
  describe('matches snapshot when', () => {
    it('has a value', () => {
      const tree = create(<ProgressBar percentage={0.4} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('has a color', () => {
      const tree = create(<ProgressBar percentage={0.4} color={Color.BRAND} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('is small', () => {
      const tree = create(<ProgressBar percentage={0.4} size={Size.SMALL} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
