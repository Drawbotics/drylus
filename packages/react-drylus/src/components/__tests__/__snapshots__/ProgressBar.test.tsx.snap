// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar matches snapshot when has a color 1`] = `
.emotion-1 {
  height: 8px;
  width: 100%;
  position: relative;
  border-radius: 8px;
  overflow: hidden;
  background: #eaeff4;
}

.emotion-1 [data-element='bar'] {
  border-radius: 8px;
}

.emotion-0 {
  position: absolute;
  z-index: 2;
  top: 0;
  left: 0;
  height: 100%;
  width: 0;
  background: #879ba6;
  -webkit-transition: all 0.15s ease-in-out;
  transition: all 0.15s ease-in-out;
  overflow: hidden;
  background: #3DB5D0;
}

.emotion-0::after {
  background: #879ba6;
  border-radius: 0 !important;
  background-image: linear-gradient( to right, #879ba6 8%, #d3dce6 18%, #879ba6 33% ) !important;
}

.emotion-0::after {
  background: #3DB5D0;
  background-image: linear-gradient( to right, #3DB5D0 8%, #bee6ef 18%, #3DB5D0 33% ) !important;
}

<div
  className="emotion-1 undefined"
>
  <div
    className="emotion-0"
    data-element="bar"
    style={
      Object {
        "width": "40%",
      }
    }
  />
</div>
`;

exports[`ProgressBar matches snapshot when has a value 1`] = `
.emotion-1 {
  height: 8px;
  width: 100%;
  position: relative;
  border-radius: 8px;
  overflow: hidden;
  background: #eaeff4;
}

.emotion-1 [data-element='bar'] {
  border-radius: 8px;
}

.emotion-0 {
  position: absolute;
  z-index: 2;
  top: 0;
  left: 0;
  height: 100%;
  width: 0;
  background: #879ba6;
  -webkit-transition: all 0.15s ease-in-out;
  transition: all 0.15s ease-in-out;
  overflow: hidden;
}

.emotion-0::after {
  background: #879ba6;
  border-radius: 0 !important;
  background-image: linear-gradient( to right, #879ba6 8%, #d3dce6 18%, #879ba6 33% ) !important;
}

<div
  className="emotion-1 undefined"
>
  <div
    className="emotion-0"
    data-element="bar"
    style={
      Object {
        "width": "40%",
      }
    }
  />
</div>
`;

exports[`ProgressBar matches snapshot when is small 1`] = `
.emotion-0 {
  position: absolute;
  z-index: 2;
  top: 0;
  left: 0;
  height: 100%;
  width: 0;
  background: #879ba6;
  -webkit-transition: all 0.15s ease-in-out;
  transition: all 0.15s ease-in-out;
  overflow: hidden;
}

.emotion-0::after {
  background: #879ba6;
  border-radius: 0 !important;
  background-image: linear-gradient( to right, #879ba6 8%, #d3dce6 18%, #879ba6 33% ) !important;
}

.emotion-1 {
  height: 8px;
  width: 100%;
  position: relative;
  border-radius: 8px;
  overflow: hidden;
  background: #eaeff4;
  height: calc(8px / 2);
  border-radius: calc(8px / 2);
}

.emotion-1 [data-element='bar'] {
  border-radius: 8px;
}

.emotion-1 [data-element='bar'] {
  border-radius: calc(8px / 2);
}

<div
  className="emotion-1"
>
  <div
    className="emotion-0"
    data-element="bar"
    style={
      Object {
        "width": "40%",
      }
    }
  />
</div>
`;
